astropy/utils/tests/test_data.py

```diff
@@ -1321,47 +1391,58 @@ def test_clear_download_cache_refuses_to_delete_outside_the_cache(tmpdir):
     assert os.path.exists(fn)
 
 
-def test_check_download_cache_finds_unreferenced_files(temp_cache, valid_urls):
+def test_check_download_cache_finds_bogus_entries(temp_cache, valid_urls):
     u, c = next(valid_urls)
     download_file(u, cache=True)
-    with _cache(pkgname='astropy', write=True) as (dldir, urlmap):
-        del urlmap[u]
-    with pytest.raises(ValueError):
+    dldir = _get_download_cache_loc()
+    bf = os.path.abspath(os.path.join(dldir, "bogus"))
+    with open(bf, "wt") as f:
+        f.write("bogus file that exists")
+    with pytest.raises(CacheDamaged) as e:
         check_download_cache()
+    assert bf in e.value.bad_files
     clear_download_cache()
 
 
-def test_check_download_cache_finds_missing_files(temp_cache, valid_urls):
+def test_check_download_cache_finds_bogus_subentries(temp_cache, valid_urls):
     u, c = next(valid_urls)
-    os.remove(download_file(u, cache=True))
-    with pytest.raises(ValueError):
+    f = download_file(u, cache=True)
+    bf = os.path.abspath(os.path.join(os.path.dirname(f), "bogus"))
+    with open(bf, "wt") as f:
+        f.write("bogus file that exists")
+    with pytest.raises(CacheDamaged) as e:
         check_download_cache()
+    assert bf in e.value.bad_files
     clear_download_cache()
 
 
-def test_check_download_cache_finds_bogus_entries(temp_cache, valid_urls):
+def test_check_download_cache_cleanup(temp_cache, valid_urls):
     u, c = next(valid_urls)
-    download_file(u, cache=True)
-    with _cache(pkgname='astropy', write=True) as (dldir, urlmap):
-        bd = os.path.join(dldir, "bogus")
-        os.mkdir(bd)
-        bf = os.path.join(bd, "file")
-        with open(bf, "wt") as f:
-            f.write("bogus file that exists")
-        urlmap[u] = bf
-    with pytest.raises(ValueError):
-        check_download_cache()
-    clear_download_cache()
+    fn = download_file(u, cache=True)
+    dldir = _get_download_cache_loc()
 
+    bf1 = os.path.abspath(os.path.join(dldir, "bogus1"))
+    with open(bf1, "wt") as f:
+        f.write("bogus file that exists")
 
-def test_check_download_cache_finds_bogus_hashes(temp_cache, valid_urls):
-    u, c = next(valid_urls)
-    fn = download_file(u, cache=True)
-    with open(fn, "w") as f:
-        f.write("bogus contents")
-    with pytest.raises(ValueError):
-        check_download_cache(check_hashes=True)
-    clear_download_cache()
+    bf2 = os.path.abspath(os.path.join(os.path.dirname(fn), "bogus2"))
+    with open(bf2, "wt") as f:
+        f.write("other bogus file that exists")
+    u2, c2 = next(valid_urls)
+    f2 = download_file(u, cache=True)
+    os.unlink(f2)
+
+    bf3 = os.path.abspath(os.path.join(dldir, "contents"))
+    with open(bf3, "wt") as f:
+        f.write("awkwardly-named bogus file that exists")
+
+    with pytest.raises(CacheDamaged) as e:
+        check_download_cache()
+    for bf in e.value.bad_files:
+        print(bf)
```

Comment:
The `print` is meant for debugging - if the test fails then `pytest` will reveal the output and you will be able to tell which files were deleted. The test is meant to ensure that `bad_files` contains a complete list in the sense that if you `clear_download_cache` all of them the implicit `check_download_cache` at the end will succeed. Asserting on their names - or just counting them - would check something different, that is, whether all the bogus files were correctly identified as bogus. Parametrized tests might be a cleaner way to handle this.

Reply:
I would still recommend rewriting it in a way that doesn't require a `print`. If you wish to show its value in case the test fails, it is clearer to have it as `assert condition, f'something was wrong, bf={bf}'`.